Cryptography API docs (#3319)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-cryptography/json-web-key-sets.md

---
type: docs
title: "JSON Web Key Sets (JWKS)"
linkTitle: "JSON Web Key Sets (JWKS)"
description: Detailed information on the JWKS cryptography component
---

## Component format

> Todo: update component format to correct format for cryptography

Reviewer: TODO here

---
type: docs

> title: "JSON Web Key Sets (JWKS)"

Reviewer 2: ```suggestion
title: "JSON Web Key Sets (JWKS) (Built-in)"
```

Author: After I push these changes, let me know if you think we still need to add to the title - I don't think we do

linkTitle: "JSON Web Key Sets (JWKS)"
description: Detailed information on the JWKS cryptography component
---

## Component format

Todo: update component format to correct format for cryptography

> 

Reviewer 2: Purpose of this component is to load keys from a JSON Web Key Set (RFC-7517). These are JSON documents that contain 1 or more keys as JWK (JSON Web Key): can be public, private, or shared keys.

This component supports loading a JWKS from a local file (if the file is changed on disk, will be reloaded automatically), from a HTTP(S) URL (will be periodically refreshed if needed), or by passing an actual JWKS in the Component YAML (as a string, which can be base64-encoded)

This is a built-in component so needs the warning.

---
type: docs
title: "JSON Web Key Sets (JWKS)"
linkTitle: "JSON Web Key Sets (JWKS)"
description: Detailed information on the JWKS cryptography component
---

## Component format

> The purpose of this component is to load keys from a JSON Web Key Set (RFC-7517). These are JSON documents that contain 1 or more keys as JWK (JSON Web Key); they can be public, private, or shared keys.

Reviewer 2: ```suggestion
The purpose of this component is to load keys from a JSON Web Key Set ([RFC 7517](https://www.rfc-editor.org/rfc/rfc7517)). These are JSON documents that contain 1 or more keys as JWK (JSON Web Key); they can be public, private, or shared keys.
```

This component supports loading a JWKS:
- From a local file (if the file is changed on disk, it's reloaded automatically),
- From a HTTP(S) URL (periodically refreshed if needed), or 

> - By passing an actual JWKS in the Component YAML (as a string, which can be base64-encoded).

Reviewer 2: ```suggestion
This component supports loading a JWKS:

- From a local file; in this case, Dapr watches for changes to the file on disk and reloads it automatically.
- From a HTTP(S) URL, which is periodically refreshed.
- By passing the actual JWKS in the `jwks` metadata property, as a string (optionally, base64-encoded).
```

{{% alert title="Note" color="primary" %}}
This component uses the cryptographic engine in Dapr to perform operations. Although keys are never exposed to your application, Dapr has access to the raw key material.

{{% /alert %}}

A Dapr `crypto.yaml` component file has the following structure:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: jwks
spec:

>   type: crypto.jwks

Reviewer 2: ```suggestion
  type: crypto.dapr.jwks
```

  version: v1
  metadata:
    - name: jwks

>       value: fixtures/crypto/jwks/jwks.json

Reviewer 2: ```suggestion
  metadata:
    # Example 1: load JWKS from file
    - name: "jwks"
      value: "fixtures/crypto/jwks/jwks.json"
    # Example 2: load JWKS from a HTTP(S) URL
    # Only "jwks" is required
    - name: "jwks"
      value: "https://example.com/.well-known/jwks.json"
    - name: "requestTimeout"
      value: "30s"
    - name: "minRefreshInterval"
      value: "10m"
    # Option 3: include the actual JWKS
    - name: "jwks"
      value: |
        {
          "keys": [
            {
              "kty": "RSA",
              "use": "sig",
              "kid": "…",
              "n": "…",
              "e": "…",
              "issuer": "https://example.com"
            }
          ]
        }
    # Option 3b: include the JWKS base64-encoded
    - name: "jwks"
      value: |
        eyJrZXlzIjpbeyJ…
```

```

{{% alert title="Warning" color="warning" %}}
The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets, as described [here]({{< ref component-secrets.md >}}).
{{% /alert %}}

## Spec metadata fields

| Field              | Required | Details | Example |
|--------------------|:--------:|---------|---------|

> | jwks               | Y        | Connection-string for the JWKS host  | `fixtures/crypto/jwks/jwks.json`
> | requestTimeout     | N        | Amount of time before request timeout; Default: 30 seconds  | `30`
> | minRefreshInterval | N        | Minimum interval for request refresh; Default: 10 minutes  | `10`

Reviewer 2: ```suggestion
| `jwks`               | Y        | Path to the JWKS document | Local file: `"fixtures/crypto/jwks/jwks.json"`<br>HTTP(S) URL: `"https://example.com/.well-known/jwks.json"`<br>Embedded JWKS: `{"keys": […]}` (can be base64-encoded)
| `requestTimeout`     | N        | Timeout for network requests when fetching the JWKS document from a HTTP(S) URL, as a Go duration. Default: "30s" | `"5s"`
| `minRefreshInterval` | N        | Minimum interval to wait before subsequent refreshes of the JWKS document from a HTTP(S) source, as a Go duration. Default: "10m" | `"1h"`
```
